File: client/src/types/gist.ts
import { User } from '@prisma/client';

export enum Visibility {
  private = 'private',
  public = 'public',
}

export enum LineWrapMode {
  noWrap = 'noWrap',
  softWrap = 'softWrap',
}

export enum IndentSize {
  two = 2,
  four = 4,
  eight = 8,
}

export type Gist = {
  id: string;
  userId: string;
  username: string;
  userImage: string;
  name: string;
  description: string;
  formula: string[];
  visibility: Visibility;
  lineWrapMode: LineWrapMode;
  indentSize: IndentSize;
  updated: string;
  created: string;
};

export type NewGist = Omit<
  Gist,
  'id' | 'username' | 'userImage' | 'updated' | 'created'
>;

export type GistDto = Omit<Gist, 'username' | 'userImage'>;

export function createGist(gist: GistDto, user: User) {
  return {
    ...gist,
    username: user.id,
    userImage: user.image ?? 'https://placehold.co/400',
  };
}

File: server/Server.API/Models/Gist.cs
namespace Server.API.Models;

class Gist
{
  [BsonId]
  [BsonRepresentation(BsonType.ObjectId)]
  public string Id { get; set; } = string.Empty;

  [BsonElement("userId")]
  public string UserId { get; set; } = string.Empty;

  [BsonElement("name")]
  public string Name { get; set; } = string.Empty;

  [BsonElement("description")]
  public string Description { get; set; } = string.Empty;

  [BsonElement("formula")]
  public List<string> Formula { get; set; } = new();

  [BsonElement("visibility")]
  public string Visibility { get; set; } = string.Empty;

  [BsonElement("lineWrapMode")]
  public string LineWrapMode { get; set; } = string.Empty;

  [BsonElement("indentSize")]
  public int IndentSize { get; set; }

  [BsonElement("created")]
  public DateTimeOffset Created { get; set; } = DateTimeOffset.Now;

  [BsonElement("updated")]
  public DateTimeOffset Updated { get; set; } = DateTimeOffset.Now;
}